security/postmortems/cve-2019-15225.md

```diff
@@ -0,0 +1,138 @@
+
+The underlying issue behind CVE-2019-15226 was first noticed via fuzzers when a timeout was reported by `h1_capture_direct_fuzz_test`: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=16325. Some profiling work revealed that `HeaderMapImpl::byteSize()`, which is O(n) in the number of headers, is called for every single header in both HTTP/1.1 and HTTP/2 codecs. Although Envoy’s stateless HTTP/2 header fuzzers (`request_header_fuzz_test` and `response_header_fuzz_test`) perform 10x more executions per second than this fuzzer, these tested one header frame per testcase and used nghttp2’s default max header frame size (16 KB). Because of this, the frame size was too small to amplify the effect of the O(n^2) process enough to produce a timeout.
+
+## Action Items
+
+* https://github.com/envoyproxy/envoy/issues/8567 
+
+## Lessons Learned
+
+### What went well
+
+* CVE-2019-15226 was detected quickly after the fuzzer reported the timeout.
+
+* The fixes for CVE-2019-15226 were straightforward and localized.
+
+* The security release occurred on time and followed the guidelines established in https://github.com/envoyproxy/envoy/blob/master/SECURITY.md
+
+### What went wrong
+
+* It took nearly a week to set up a branch for fix patches. This was due to some confusion over whether to use the new GitHub Security advisories, which didn’t support the required permission model and CI integrations. In the process, the envoy-setec branch was temporarily made readable to all Envoy contributors. 
+
+* While resolving the above permission issue, we hit an issue with Github permissions on envoyproxy: people could no longer assign issues to members in the Envoy repository. This was fixed with some restructuring of GitHub team’s to support the limited GitHub IAM model.
+
+* It was possible to push to envoy-setec branches by fix team, e.g. the 1.11.2 could be directly pushed to (master as well). We need branch protection to ensure that CI gates merges; this will provide confidence that the staged release branches are likely to work on the main Envoy repository.
+
+* We had manual patch sets the day before release, but no envoy-setec branches reflecting them passing end-to-end. We should not consider a release ready to go until it passes a full CI pass.
+
+* It wasn’t possible to get a full CI pass due to docs/image/etc push issues. We should have a set of presubmits that provide a simple yes/no in the GH UX.
+
+* Our route resolution fuzzer would not have picked up the regex vulnerability due to a logical error in the fuzzer.
```

Comment:
I think one AI we can take from this is that new fuzzers (or additions to existing fuzzers) should come with gtest-style unit tests, with some way to validate that sample inputs do something useful.

Reply:
Added